resources/hostname.go

```diff
@@ -0,0 +1,259 @@
+			Name: name,
+		},
+		Hostname: hostname,
+	}
+	return obj, obj.Init()
+}
+
+// Init runs some startup code for this resource.
+func (obj *HostnameRes) Init() error {
+	obj.BaseRes.kind = "Hostname"
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Validate if the params passed in are valid data.
+// FIXME: where should this get called ?
+func (obj *HostnameRes) Validate() error {
+	return nil
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *HostnameRes) Watch(processChan chan event.Event) error {
+	if obj.IsWatching() {
+		return nil // TODO: should this be an error?
+	}
+	obj.SetWatching(true)
+	defer obj.SetWatching(false)
+	cuid := obj.converger.Register()
+	defer cuid.Unregister()
+
+	var startup bool
+	Startup := func(block bool) <-chan time.Time {
+		if block {
+			return nil // blocks forever
+			//return make(chan time.Time) // blocks forever
+		}
+		return time.After(time.Duration(500) * time.Millisecond) // 1/2 the resolution of converged timeout
+	}
+
+	// if we share the bus with others, we will get each others messages!!
+	bus, err := util.SystemBusPrivateUsable() // don't share the bus connection!
+	if err != nil {
+		return errwrap.Wrap(err, "Failed to connect to bus")
+	}
+	defer bus.Close()
+	bus.BusObject().Call("org.freedesktop.DBus.AddMatch", 0,
+		"type='signal',path='/org/freedesktop/hostname1',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged'")
+	signals := make(chan *dbus.Signal, 10)
+	bus.Signal(signals)
+
+	var send = false // send event?
+	var exit = false
+	var dirty = false
+	var closed = false
+
+	defer func() {
+		if !closed {
+			close(signals)
+		}
+	}()
+
+	for {
+		obj.SetState(ResStateWatching) // reset
+		select {
+		case event := <-signals:
+			cuid.SetConverged(false)
+			if event == nil {
+				closed = true
+			}
+			send = true
+			dirty = true
+		case event := <-obj.Events():
+			cuid.SetConverged(false)
+			// we avoid sending events on unpause
+			if exit, send = obj.ReadEvent(&event); exit {
+				return nil // exit
+			}
+			send = true
+			dirty = true
+
+		case <-cuid.ConvergedTimer():
+			cuid.SetConverged(true) // converged!
+			continue
+
+		case <-Startup(startup):
+			cuid.SetConverged(false)
+			send = true
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			startup = true // startup finished
+			send = false
+			if dirty {
+				dirty = false
+				obj.isStateOK = false // something made state dirty
+			}
+			// only do this on certain types of events
+			if exit, err := obj.DoSend(processChan, ""); exit || err != nil {
+				return err // we exit or bubble up a NACK...
+			}
+		}
+	}
+}
+
+func (obj *HostnameRes) updateProperty(object dbus.BusObject, property, setterName string) (bool, error) {
+	if hostnameProperty, err := object.GetProperty("org.freedesktop.hostname1." + property); err == nil {
+		switch hostname := hostnameProperty.Value().(type) {
```

Comment:
Instead of a switch like this, we can inline it all I think, eg:
https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

IOW: if err != nil: return
if !string return
etc...